### PR TITLE
[model] feat: Support qwen3-omni-moe model

### PR DESCRIPTION
**Pull request by Crystal-jiang**
### What does this PR do?

> Add support for the `qwen3-omni-moe` model, including necessary configuration and adapter integration.
> Modified on PR  https://github.com/ByteDance-Seed/VeOmni/pull/422
> Related to Issue https://github.com/ByteDance-Seed/VeOmni/issues/393

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `misc`, `ci`, `config`, `docs`, `data`, `dist`, `omni`, `logging`, `model`, `optim`, `ckpt`, `release`, `task`, `perf`, `ops`, `parallel`
  - If this PR involves multiple modules, separate them with `,` like `[ci, data, model]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by Crystal-jiang** [comment]
Plan to refactor the modeling part into patch-based implementation.

**Comment by Luosuu** [comment]
let's have some e2e smoke tests so that we are sure it is runnable at least

**Comment by Crystal-jiang** [comment]
> let's have some e2e smoke tests so that we are sure it is runnable at least

I’m currently adding the e2e smoke test scripts. Note that qwen3-omni-moe is a 30B model, so running full e2e tests in CI could take considerable time. I’ll finalize the scripts and then coordinate with the community on model and data preparation.

**Comment by Luosuu** [comment]
we can add toy model config like other models so that we are not running the real 30B model.

File: tests/e2e/exec_scripts.py [hunk 1]
```diff
@@ -24,8 +24,26 @@ def qwen3_0p6b_base_tulu_sft_script():
     return exec_script
 
 
+def qwen3_omni_moe_coco_sft_script():
+    params = [
+        "torchrun --nnodes=1 --nproc_per_node=8 --master-port=4322",
+        "tasks/omni/train_qwen3_omni.py",
+        "configs/multimodal/qwen3_omni/qwen3_omni.yaml",
+        f"--model.model_path {os.path.dirname(os.path.abspath(__file__)) + '/../models/toy_config/qwen3omni_toy'}",
+        f"--data.train_path {os.path.join(CI_DATASET_DIR, 'sharegpt4v_instruct_gpt4-vision_cap100k_coco.json')}",
```
Comment (pjgao) [inline tests/e2e/exec_scripts.py]: Can we add toy data under tests/data?

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py
Comment (TimYangst) [inline veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py:1]: Could we merge pure huggingface qwen3_omni_moe model at first as review baseline? Such that it is more clearly to check which parts are modified.

cc @Luosuu @Coach257  @FoolPlayer

File: veomni/models/transformers/qwen3_omni_moe/configuration_qwen3_omni_moe.py
Comment (TimYangst) [inline veomni/models/transformers/qwen3_omni_moe/configuration_qwen3_omni_moe.py:1]: Do we modify the configuration file here? If not (which seems likely), we could directly import the configuration class from Transformers instead:
`transformers.models.qwen3_omni_moe.configuration_qwen3_omni_moe`

File: veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py
Comment (TimYangst) [inline veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py:1]: Same for this file, if it’s not modified, we can import it directly from Transformers.

File: veomni/models/transformers/qwen3_omni_moe/configuration_qwen3_omni_moe.py
Comment (Crystal-jiang) [reply veomni/models/transformers/qwen3_omni_moe/configuration_qwen3_omni_moe.py:1]: Both of these files(configuration and processing) have minor modifications and cannot be used directly via import.

File: veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py
Comment (Crystal-jiang) [reply veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py:1]: Both of these files(configuration and processing) have minor modifications and cannot be used directly via import.

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py
Comment (Crystal-jiang) [reply veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py:1]: Sure, I'll first submit a PR for the pure HuggingFace version of the Qwen3-Omni-MoE model. This new PR will include all files without any modifications. Once the new PR is merged, I'll rebase this current PR #409 to clearly highlight the changes.

File: veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py
Comment (TimYangst) [reply veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py:1]: It looks like the changes in this file are semantically identical. Are they necessary?

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py [hunk 2]
```diff
@@ -1480,6 +1578,7 @@ class Qwen3OmniMoeThinkerTextDecoderLayer(GradientCheckpointingLayer):
     def __init__(self, config, layer_idx):
         super().__init__()
         self.hidden_size = config.hidden_size
+
```
Comment (TimYangst) [inline veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: nits: remove this empty line

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py [hunk 3]
```diff
@@ -2063,10 +2228,29 @@ def forward(
...
+        if self.training and get_parallel_state().sp_enabled:
+            # (batch_size, seq_len // sp_size, hidden_size) to  (batch_size, seq_len, hidden_size // sp_size)
+            inputs_embeds = gather_seq_scatter_heads(
+                inputs_embeds, seq_dim=1, head_dim=2, group=get_parallel_state().sp_group
+            )
+
         # 2. Merge text , audios , image and video
         if input_features is not None:
+            # TODO audio
```
Comment (TimYangst) [inline veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: Maybe we can add a bit more explicit comments about the remaining work here.
In particular, the audio path does not seem to be SP-compatible yet (unlike image/video).
It might also be safer to explicitly raise an error when audio inputs are provided while SP is enabled, to avoid silently incorrect behavior.

**Review by TimYangst — APPROVED** [review]

File: veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py
Comment (Crystal-jiang) [reply veomni/models/transformers/qwen3_omni_moe/processing_qwen3_omni_moe.py:1]: Removed the `is not None` checks at three locations (lines 173, 187, 194) in the processing file. Take line 187: `if images is not None`: as an example. When training with multisource data, `images` is an empty list [], so this check evaluates to `True`, which triggers errors when fetching values from `images` and terminates training. This modification is therefore necessary.

**Comment by Luosuu** [comment]
Hi @Crystal-jiang, feel free to merge when you feel it is ready. LGTM.
